### PR TITLE
Add m_stride to layout to support propagation of Padding

### LINKED ISSUE #2995: Mirror not getting correct stride with Padding enabled.
If padding is requested, the stride is not correctly added to the mirror. 
```c++
#include<Kokkos_Core.hpp>
#include<cmath>

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv);
  {
    int N = argc>1?atoi(argv[1]):11;
    int R = argc>2?atoi(argv[2]):7;

    auto alloc = Kokkos::view_alloc("test", Kokkos::AllowPadding);
    auto darr = Kokkos::View<double**, Kokkos::CudaSpace>(alloc, N, R);
    auto harr = Kokkos::create_mirror_view(darr);

    std::cout << "darr: " << darr.stride(0) << ", " << darr.stride(1) << "\n";
    std::cout << "harr: " << harr.stride(0) << ", " << harr.stride(1) << "\n";

    Kokkos::deep_copy(harr, darr);

  }
  Kokkos::finalize();
}
```

### PR DESCRIPTION
**Pull request by janciesko**
Propagate padding when `create_mirror_view` and source view has `Kokkos::AllowPadding`.

Case in point:
```
auto alloc = Kokkos::view_alloc("test", Kokkos::AllowPadding);
auto v = Kokkos::View<double**, Kokkos::CudaSpace>(alloc, 17, 4);
auto h_v = Kokkos::create_mirror_view(v);
ASSERT_EQ(v.stride(1), h_v.stride(1));
```

Fixes #2995

### CODE REVIEW

**Comment by janciesko** [comment]
Todo: Unit test coverage

**Comment by janciesko** [comment]
This fails in:
cuda.view_layouttiled
unknown file: Failure
C++ exception with description "Deprecation Error: Kokkos::deep_copy extents of views don't match: dv(4,12,16,12) dv_mirror(4,12,16,16) " thrown in the test body.

This is understandable. We need to fix this by supporting deep copies between views with padding but where padding is coming from View type (Kokkos::AllowPadding) and not subview.

**Comment by ajpowelsnl** [comment]
Hi @janciesko - do you know the path forward for this issue?

**Comment by janciesko** [comment]
@ajpowelsnl  - Yep, looking at it now.

**Review by masterleinad — DISMISSED** [review]
Does this also work with the other View-like types (`DualView`, `DynamicView`, `DynRankView`, `OffsetView`)?

Add some unit tests.

**Comment by PhilMiller** [comment]
Here's a weird edge case to consider - should we accept copies from a strided view to a suitably padded view, such that a contiguous `memcpy` would not overwrite any valid data?

If someone called for mirror creation of a strided subview, with suitable options to allow padding, this could be a path to facilitating it

**Comment by PhilMiller** [comment]
> Does this also work with the other View-like types (`DualView`, `DynamicView`, `DynRankView`, `OffsetView`)?

`DynRankView` should be happy with whatever we do for `View`, since `deep_copy` calls on it just check that ranks match, and then call down to the corresponding overload for the corresponding static-rank `View` type.

`DynamicView` only exists for rank-1 anyway, and never actually provides contiguous layout for the data, so it's really not relevant.

`OffsetView` is still under `Experimental`, so I think it's fair not to support it with the full range of routines.

`DualView` is a monster

File: containers/src/Kokkos_OffsetView.hpp [hunk 1]
```diff
@@ -1094,6 +1094,43 @@ class OffsetView : public ViewTraits<DataType, Properties...> {
     return m_track.template get_label<typename traits::memory_space>();
   }
 
+
+  private:
+  template <class T, int R,
+            bool P = (R > 1) && (std::is_same<T, Kokkos::LayoutLeft>::value ||
+                                 std::is_same<T, Kokkos::LayoutRight>::value)>
+  struct is_strided_from_padding;
```
Comment (PhilMiller) [inline containers/src/Kokkos_OffsetView.hpp:1102]: Rather than repeat this helper class as a private member of each `View`-like class, could it just go somewhere in `Kokkos::Impl`?

File: containers/src/Kokkos_DynRankView.hpp [hunk 2]
```diff
@@ -1088,6 +1088,42 @@ class DynRankView : public ViewTraits<DataType, Properties...> {
...
+    is_strided_from_padding(const bool& arg){};
+    KOKKOS_INLINE_FUNCTION
+    bool value() const { return false; }
+  };
+
+  template <class T, int R>
+  struct is_strided_from_padding<T, R, true> {
+    bool v;
+    KOKKOS_INLINE_FUNCTION
+    is_strided_from_padding() = default;
```
Comment (masterleinad) [inline containers/src/Kokkos_DynRankView.hpp:1111]: What's the default value for `v`?

File: containers/src/Kokkos_DynRankView.hpp [hunk 3]
```diff
@@ -2010,7 +2049,8 @@ inline auto create_mirror(
   static_cast<Impl::ViewCtorProp<void, std::string>&>(prop_copy).value =
       std::string(src.label()).append("_mirror");
 
-  return dst_type(prop_copy, Impl::reconstructLayout(src.layout(), src.rank()));
+  return dst_type(prop_copy, Impl::reconstructLayout(src.layout(), src.rank()),
+                  src.impl_is_strided_from_padding());
```
Comment (masterleinad) [inline containers/src/Kokkos_DynRankView.hpp:2053]: Just add `src.impl_is_strided_from_padding()` to `prop_copy` here and in the other places.

File: containers/src/Kokkos_OffsetView.hpp [hunk 4]
```diff
@@ -1182,9 +1219,11 @@ class OffsetView : public ViewTraits<DataType, Properties...> {
       const Kokkos::Impl::ViewCtorProp<P...>& arg_prop,
       std::enable_if_t<!Kokkos::Impl::ViewCtorProp<P...>::has_pointer,
                        typename traits::array_layout> const& arg_layout,
-      const index_list_type minIndices)
+      const index_list_type minIndices, bool strided_from_padding   = false)
       : m_track(),
-        m_map()
+        m_map(),
+         m_is_strided_from_padding(Kokkos::Impl::ViewCtorProp<P...>::allow_padding ||
+                                  strided_from_padding)
```
Comment (masterleinad) [inline containers/src/Kokkos_OffsetView.hpp:1226]: ... or is there any problem if arg_prop allows padding in case strided_from_padding is passed?

Comment (janciesko) [reply containers/src/Kokkos_OffsetView.hpp]: That is a good point, ViewCtorProp is a better location for the member.

File: containers/src/Kokkos_DynRankView.hpp (hunk 3 above)
Comment (janciesko) [reply containers/src/Kokkos_DynRankView.hpp:2030]: Ok

File: containers/src/Kokkos_DynRankView.hpp (hunk 2 above)
Comment (janciesko) [reply containers/src/Kokkos_DynRankView.hpp]: false

File: containers/src/Kokkos_OffsetView.hpp (hunk 1 above)
Comment (janciesko) [reply containers/src/Kokkos_OffsetView.hpp]: Yes.

**Comment by janciesko** [comment]
Merge https://github.com/kokkos/kokkos/pull/5314 into develop, rebase and use the following to set the state. Remove other changes.
```
if (src.allow_paddding)
  Impl::add_properties(prop_copy, AllowPadding_t()); 
```

**Comment by masterleinad** [comment]
> Wait for https://github.com/kokkos/kokkos/pull/5314, rebase and use the following to set the state. Remove other changes.
> ```
> if (src.allow_paddding)
>   Impl::add_properties(prop_copy, AllowPadding_t());
> ```

@janciesko ping

File: core/src/Kokkos_View.hpp [hunk 5]
```diff
@@ -123,6 +123,32 @@ void runtime_check_rank(const size_t rank, const size_t dyn_rank,
...
+  bool value() const { return false; }
+};
+
+template <class T, int R>
+struct is_strided_from_padding<T, R, true> {
+  bool v = false;
+  KOKKOS_INLINE_FUNCTION
+  is_strided_from_padding() = default;
+  KOKKOS_INLINE_FUNCTION
+  is_strided_from_padding(const bool& arg) { v = arg; };
```
Comment (PhilMiller) [inline core/src/Kokkos_View.hpp:147]:
```suggestion
  is_strided_from_padding(const bool& arg) : v(arg) { };
```

File: core/src/Kokkos_CopyViews.hpp [hunk 6]
```diff
@@ -1776,7 +1776,9 @@ inline void deep_copy(
       (std::is_same<typename dst_type::array_layout,
                     typename src_type::array_layout>::value ||
        (dst_type::rank == 1 && src_type::rank == 1)) &&
-      dst.span_is_contiguous() && src.span_is_contiguous() &&
```
Comment (PhilMiller) [inline core/src/Kokkos_CopyViews.hpp:1779]: The comment above needs to be amended to match

Comment (PhilMiller) [reply core/src/Kokkos_CopyViews.hpp:1779]: And similarly above the next change as well

**Review by PhilMiller — COMMENTED** [review]
There should probably be a corresponding unit test for this

**Review by crtrott — CHANGES_REQUESTED** [review]
We need a test to make sure this now works.

File: core/src/Kokkos_CopyViews.hpp [hunk 7]
```diff
@@ -3492,19 +3496,14 @@ create_mirror(const Kokkos::View<T, P...>& src,
       "The view constructor arguments passed to Kokkos::create_mirror must "
       "not explicitly allow padding!");
 
+  if (src.impl_is_strided_from_padding()) {
+    auto prop_copy = Impl::with_properties_if_unset(
+        arg_prop, std::string(src.label()).append("_mirror"), AllowPadding);
```
Comment (PhilMiller) [inline core/src/Kokkos_CopyViews.hpp:3501]: Could you factor out the common expression for the label string?

File: core/src/Kokkos_CopyViews.hpp [hunk 8]
```diff
@@ -3532,30 +3531,14 @@ create_mirror(const Kokkos::View<T, P...>& src,
...
-  layout.stride[4] = src.stride_4();
-  layout.stride[5] = src.stride_5();
-  layout.stride[6] = src.stride_6();
-  layout.stride[7] = src.stride_7();
-
+  if (src.impl_is_strided_from_padding()) {
+    auto prop_copy = Impl::with_properties_if_unset(
+        arg_prop, std::string(src.label()).append("_mirror"), AllowPadding);
+    return dst_type(prop_copy, src.layout());
+  }
```
Comment (masterleinad) [inline core/src/Kokkos_CopyViews.hpp:3538]: You don't need the specialization on the Layout anymore, do you?

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 9]
```diff
@@ -721,6 +721,37 @@ struct SubviewExtents {
...
+  KOKKOS_INLINE_FUNCTION
+  static constexpr size_t stride(size_t const N) {
+    return ((align != 0) &&
+            ((static_cast<int>(Kokkos::Impl::MEMORY_ALIGNMENT_THRESHOLD) *
+              static_cast<int>(align)) < N) &&
+            ((N % div_ok) != 0))
+               ? N + align - (N % div_ok)
+               : N;
+  }
+};
```
Comment (masterleinad) [inline core/src/impl/Kokkos_ViewMapping.hpp:753]: I find this really hard to read. Maybe more in the direction of
```suggestion
template <unsigned TrivialScalarSize>
struct Padding {
  // If memory alignment is a multiple of the trivial scalar size then attempt
  // to align.
  KOKKOS_INLINE_FUNCTION
  static constexpr size_t stride(size_t const N) {
    if constexpr(TrivialScalarSize ==0) {
      return N;
    } else {
      constexpr int div  = Kokkos::Impl::MEMORY_ALIGNMENT / TrivialScalarSize;
      constexpr int mod = Kokkos::Impl::MEMORY_ALIGNMENT / TrivialScalarSize;
      constexpr int align = (mod == 0) ? div : 0;
      // To valid modulo zero in constexpr
      constexpr int div_ok = (div != 0) ? div : 1;
      if constexpr ((align != 0) &&
          ((static_cast<int>(Kokkos::Impl::MEMORY_ALIGNMENT_THRESHOLD) *
              static_cast<int>(align)) < N) &&
            ((N % div_ok) != 0))
        return N + align - (N % div_ok);
      else
        return N;
    }
  }
};
```

File: core/unit_test/TestViewPadding.hpp [hunk 10]
```diff
@@ -0,0 +1,564 @@
...
+                             use_padding::value ? sizeof(DType) : 0>;
+  offset_t offset = offset_t(padding(), layout(params...));
+  auto alloc_prop = Kokkos::view_alloc("vDim" + std::to_string(dim));
+  view_t v(alloc_prop, params...);
+  return ASSERT_OFFSET_STRIDES(v, offset) ? true : false;
+}
+
+template <typename T, typename viewDim, typename layout, typename allow_padding,
+          typename... Vals>
+bool padding_assignement_and_copy(int dim, Vals... params) {
```
Comment (masterleinad) [inline core/unit_test/TestViewPadding.hpp:150]:
```suggestion
bool padding_assignment_and_copy(int dim, Vals... params) {
```

File: core/src/Kokkos_CopyViews.hpp (hunk 7 above)
Comment (janciesko) [reply core/src/Kokkos_CopyViews.hpp:3501]: Yes, I'll force push

File: core/src/Kokkos_View.hpp [hunk 11]
```diff
@@ -123,6 +123,32 @@ void runtime_check_rank(const size_t rank, const size_t dyn_rank,
   }
 }
 
+template <class T, int R,
+          bool P = (R > 1) && (std::is_same<T, Kokkos::LayoutLeft>::value ||
+                               std::is_same<T, Kokkos::LayoutRight>::value)>
```
Comment (PhilMiller) [inline core/src/Kokkos_View.hpp:128]:
```suggestion
          bool P = (R > 1) && (std::is_same_v<T, Kokkos::LayoutLeft> ||
                               std::is_same_v<T, Kokkos::LayoutRight>)>
```
Since I don't think we intend to backport this to 3.7.x, we can use the nicer C++17 spelling.

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 12]
```diff
@@ -3430,8 +3386,15 @@ class ViewMapping<
     // Query the mapping for byte-size of allocation.
     // If padding is allowed then pass in sizeof value type
     // for padding computation.
-    using padding = std::integral_constant<
-        unsigned int, alloc_prop::allow_padding ? sizeof(value_type) : 0>;
+
+    using use_padding =
+        std::bool_constant<alloc_prop::allow_padding &&
+                           !std::is_same<typename Traits::array_layout,
+                                         Kokkos::LayoutStride>::value>;
```
Comment (PhilMiller) [inline core/src/impl/Kokkos_ViewMapping.hpp:3393]:
```suggestion
                           !std::is_same_v<typename Traits::array_layout,
                                         Kokkos::LayoutStride>>;
```

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 13]
```diff
@@ -721,6 +721,32 @@ struct SubviewExtents {
...
+struct Padding {
+  // If memory alignment is a multiple of the trivial scalar size then attempt
+  // to align.
+  KOKKOS_INLINE_FUNCTION
+  static constexpr size_t stride(size_t const N) {
+    if constexpr (TrivialScalarSize == 0) {
+      return N;
+    } else {
+      constexpr int div   = Kokkos::Impl::MEMORY_ALIGNMENT / TrivialScalarSize;
+      constexpr int mod   = Kokkos::Impl::MEMORY_ALIGNMENT / TrivialScalarSize;
```
Comment (PhilMiller) [inline core/src/impl/Kokkos_ViewMapping.hpp:734]:
```suggestion
      constexpr int mod   = Kokkos::Impl::MEMORY_ALIGNMENT % TrivialScalarSize;
```

**Review by PhilMiller — CHANGES_REQUESTED** [review]
That looks like a mistake

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 13 above)
Comment (PhilMiller) [reply core/src/impl/Kokkos_ViewMapping.hpp:734]: Also, if the tests didn't/don't catch that, then it exposes a gap in the tests

Comment (janciesko) [reply core/src/impl/Kokkos_ViewMapping.hpp:734]: Yes, I'll add a check that that Padding has effect.

File: core/src/Kokkos_View.hpp (hunk 11 above)
Comment (janciesko) [reply core/src/Kokkos_View.hpp:128]: Thanks, yep

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 12 above)
Comment (janciesko) [reply core/src/impl/Kokkos_ViewMapping.hpp:3392]: We should update this across files (Kokkos_View, Kokkos_ViewMapping, etc.) so I will defer this to another PR.

**Review by PhilMiller — COMMENTED** [review]
Error got fixed, and my other comments got addressed. I don't have time to review in full, but I will unblock things for others to approve

**Comment by janciesko** [comment]
I've updated the unit test, it reflects the types when we pad and mirror. Note: subviews that remain contiguous, that is, subviews that slice among the outer dims for the given layout, do not inherit padding. The other use case where non-contagiousness is coming from padding is supported now and a deep_copy succeeds.

File: core/unit_test/TestViewPadding.hpp [hunk 14]
```diff
@@ -0,0 +1,767 @@
...
+#include <sstream>
+#include <iostream>
+#include <time.h>
+#include <string>
+
+#include <Kokkos_Core.hpp>
+
+namespace Test {
+
+#define UNUSED(x) (void)(x)
```
Comment (PhilMiller) [inline core/unit_test/TestViewPadding.hpp:56]: Relevant variables can be declared with `[[maybe_unused]]` now

Comment (PhilMiller) [reply core/unit_test/TestViewPadding.hpp:56]: Or since this is only applied to function parameters, and not local variables, could they just be left unnamed?

File: core/unit_test/TestViewPadding.hpp [hunk 15]
```diff
@@ -0,0 +1,767 @@
...
+#undef PARAM_4_RANK
+#undef PARAM_5_RANK
+#undef PARAM_6_RANK
+#undef PARAM_7_RANK
+
+#undef SUBRANGE_DIM
+
+#undef UNUSED
+
+#undef DType
```
Comment (PhilMiller) [inline core/unit_test/TestViewPadding.hpp:765]: This is not a defined macro.

There are missing undefines for `ZERO` and `ARG_*`

File: core/src/Kokkos_CopyViews.hpp [hunk 16]
```diff
@@ -1776,7 +1776,9 @@ inline void deep_copy(
       (std::is_same<typename dst_type::array_layout,
                     typename src_type::array_layout>::value ||
        (dst_type::rank == 1 && src_type::rank == 1)) &&
-      dst.span_is_contiguous() && src.span_is_contiguous() &&
+      ((dst.span_is_contiguous() && src.span_is_contiguous()) ||
+       (dst.impl_is_strided_from_padding() &&
+        src.impl_is_strided_from_padding())) &&
```
Comment (PhilMiller) [inline core/src/Kokkos_CopyViews.hpp:1781]: The comments at 1771 and 2923 still need to be updated

**Comment by PhilMiller** [comment]
Earlier versions of this PR explicitly modified `Kokkos_DynRankView.hpp` as well - is that no longer necessary? Could we get some testing to make sure that it's covered too?

**Comment by janciesko** [comment]
We need to add that member to some of the containers. I am adding another test to test padding propagation for containers.

**Comment by janciesko** [comment]
Upon closer inspection, the only relevant container type is DualView. I have added the unit test.

**Comment by janciesko** [comment]
Checking "/__w/kokkos/kokkos/core/unit_test/TestViewPadding.hpp:587: Failure"

**Comment by PhilMiller** [comment]
Both failures in Jenkins now are actually timing-based, and unrelated to this PR

File: containers/unit_tests/TestPaddingPropagation.hpp [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+using DType_1 = DType *;
+using DType_2 = DType **;
+using DType_3 = DType ***;
+using DType_4 = DType ****;
+using DType_5 = DType *****;
+using DType_6 = DType ******;
+using DType_7 = DType *******;
+
+#define ASSERT_VIEW_STRIDES(view1, view2, rank)           \
+  (view1.stride(0) == view2.stride(0) || rank < 1) &&     \
```
Comment (PhilMiller) [inline containers/unit_tests/TestPaddingPropagation.hpp:83]: For clarity of control flow, I'd rather each line's condition were written in the opposite order, 
```suggestion
  (rank < 1 || view1.stride(0) == view2.stride(0)) &&     \
```
But I don't see that as a gating condition.

File: containers/unit_tests/TestPaddingPropagation.hpp [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+#undef PARAM_0
+#undef PARAM_1
+#undef PARAM_2
+#undef PARAM_3
+#undef PARAM_4
+#undef PARAM_5
+#undef PARAM_6
+#undef PARAM_7
+
+#undef VIEW_T
```
Comment (PhilMiller) [inline containers/unit_tests/TestPaddingPropagation.hpp:151]:
```suggestion
#undef VIEW_T
#undef ASSERT_VIEW_STRIDES
```

File: containers/unit_tests/TestPaddingPropagation.hpp [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+      (view1.stride(3) == view2.stride(3) || rank < 4) && \
+      (view1.stride(4) == view2.stride(4) || rank < 5) && \
+      (view1.stride(5) == view2.stride(5) || rank < 6) && \
+      (view1.stride(6) == view2.stride(6) || rank < 7) && \
+      (view1.stride(7) == view2.stride(7) || rank < 8)
+
+template <typename dt, typename layout, typename... Vals>
+bool padding_assignment_and_copy(int rank, Vals... params) {
+  bool result      = true;
+  using DualView_t = VIEW_T<dt, layout>;
```
Comment (PhilMiller) [inline containers/unit_tests/TestPaddingPropagation.hpp:95]: The `VIEW_T` macro used here seems pointless, since it's only used on this one line, and the rest of the code assumes it's `DualView` and not anything else.
```suggestion
  using DualView_t = Kokkos::DualView<dt, layout>;
```

File: containers/unit_tests/TestPaddingPropagation.hpp [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+//@HEADER
+*/
+
+#include <gtest/gtest.h>
+
+#include <Kokkos_DualView.hpp>
+#include <Kokkos_Core.hpp>
+
+#define VIEW_T Kokkos::DualView
+
```
Comment (PhilMiller) [inline containers/unit_tests/TestPaddingPropagation.hpp:51]:
```suggestion
```

**Review by PhilMiller — COMMENTED** [review]
If you wanted to do away with the `PARAM_N` macros in the tests, C++17 `std::apply` would help out:
https://stackoverflow.com/questions/687490/how-do-i-expand-a-tuple-into-variadic-template-functions-arguments
I'd bet on that making nvcc or some other vendor compiler unhappy, though, so I'm fine with leaving it as is.

File: core/unit_test/TestViewPadding.hpp [hunk 21]
```diff
@@ -0,0 +1,618 @@
...
+padding_subview_and_copy(int rank, Vals... params) {
+  bool result  = true;
+  using view_t = Kokkos::View<dim_t, layout>;
+  auto alloc_prop =
+      Kokkos::view_alloc("vDim" + std::to_string(rank), Kokkos::AllowPadding);
+  view_t v(alloc_prop, params...);
+  auto v_sub        = get_subview<dim_t, layout, viewDim>(v);
+  auto v_sub_mirror = Kokkos::create_mirror_view(v_sub);
+  result &= ASSERT_VIEW_STRIDES(v, v_sub, rank);
+  if (v_sub_mirror.span_is_contiguous())
```
Comment (PhilMiller) [inline core/unit_test/TestViewPadding.hpp:342]: Do we need to test that `v_sub.span_is_contiguous() == v_sub_mirror.span_is_contiguous()`? It seems like relying on the mirror having that correct is an unjustified assumption about the code exercised by the test.

File: core/unit_test/TestViewPadding.hpp [hunk 22]
```diff
@@ -0,0 +1,618 @@
...
+padding_subview_and_copy(int rank, Vals... params) {
+  bool result  = true;
+  using view_t = Kokkos::View<dim_t, layout>;
+  auto alloc_prop =
+      Kokkos::view_alloc("vDim" + std::to_string(rank), Kokkos::AllowPadding);
+  view_t v(alloc_prop, params...);
+  auto v_sub        = get_subview<dim_t, layout, viewDim>(v);
+  auto v_sub_mirror = Kokkos::create_mirror_view(v_sub);
+  result &= ASSERT_VIEW_STRIDES(v, v_sub, rank);
+  /* We never pad for rank 0 and rank 1 and for LR && dyn_rank > 1 */
```
Comment (PhilMiller) [inline core/unit_test/TestViewPadding.hpp:322]: The comment mentions `LayoutRight && dyn_rank > 1`, but the code doesn't make such a distinction - it only considers rank.

File: core/unit_test/TestViewPadding.hpp [hunk 23]
```diff
@@ -0,0 +1,618 @@
...
+
+template <typename dim_t, typename layout, typename viewDim>
+auto get_subview(
+    Kokkos::View<dim_t, layout> &view,
+    [[maybe_unused]] typename std::enable_if_t<
+        std::is_same<viewDim, Kokkos::Impl::ViewDimension<ARG_1>>::value, void>
+        *ptr = nullptr) {
+  auto pair = pair_t(SUBRANGE_DIM);
+  return Kokkos::subview(view, pair);
+}
```
Comment (PhilMiller) [inline core/unit_test/TestViewPadding.hpp:242]: The `get_subview` overloads can almost certainly be consolidated down to one case. If nothing else, the body can just be a ladder of `if constexpr` on the same conditions, but I think it can actually be written generically. I'll only push on going at least as far as `if constexpr` though.

**Review by PhilMiller — CHANGES_REQUESTED** [review]
Could you cover `DynRankView` in the unit tests too?

Otherwise, I think this looks good.

File: core/unit_test/TestViewPadding.hpp (hunk 21 above)
Comment (janciesko) [reply core/unit_test/TestViewPadding.hpp:342]: We are checking here that a mirror of a subview that is contiguous in memory has different strides as a result of a different size. This must happen if the subview is a slice among the first significant dimension. We do this to guarantee that the addition of padding still allows contiguous subwies.

**Comment by janciesko** [comment]
Sure, adding the test. For DynRankView we do not use padding, so we can test if a padded view propagates correctly in that corresponding constructor.

File: core/unit_test/TestViewPadding.hpp (hunk 23 above)
Comment (janciesko) [reply core/unit_test/TestViewPadding.hpp:242]: No, I think that wouldn't work as the return value type is different.

**Comment by janciesko** [comment]
Test added.

**Review by PhilMiller — APPROVED** [review]

File: core/src/Kokkos_CopyViews.hpp [hunk 24]
```diff
@@ -3492,70 +3493,15 @@ create_mirror(const Kokkos::View<T, P...>& src,
...
-      arg_prop, std::string(src.label()).append("_mirror"));
-
-  return dst_type(prop_copy, layout);
+  if (src.impl_is_strided_from_padding()) {
+    auto prop_copy =
+        Impl::with_properties_if_unset(arg_prop, label, AllowPadding);
+    return dst_type(prop_copy, src.layout());
+  }
+  auto prop_copy = Impl::with_properties_if_unset(arg_prop, label);
+  return dst_type(prop_copy, src.layout());
```
Comment (dalg24) [inline core/src/Kokkos_CopyViews.hpp:3504]: Did you mean to drop the `_mirror` suffix?

Comment (dalg24) [reply core/src/Kokkos_CopyViews.hpp]: Nevermind I see it is appended to `label`.  Sorry about the noise.

**Comment by PhilMiller** [comment]
The merge looks good to me

File: core/src/Kokkos_View.hpp [hunk 25]
```diff
@@ -123,6 +123,32 @@ void runtime_check_rank(const size_t rank, const size_t dyn_rank,
   }
 }
 
+template <class T, int R,
+          bool P = (R > 1) && (std::is_same<T, Kokkos::LayoutLeft>::value ||
```
[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/Kokkos_View.hpp:127]: P is the condition on which we specialize. Renamed to Condition

File: core/src/Kokkos_View.hpp [hunk 26]
```diff
@@ -123,6 +123,32 @@ void runtime_check_rank(const size_t rank, const size_t dyn_rank,
...
+                               std::is_same<T, Kokkos::LayoutRight>::value)>
+struct is_strided_from_padding;
+
+template <class T, int R>
+struct is_strided_from_padding<T, R, false> {
+  KOKKOS_INLINE_FUNCTION
+  is_strided_from_padding() = default;
```
[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/Kokkos_View.hpp:134]: We need to use the compiler generated (default) function here.

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 27]
```diff
@@ -721,6 +721,31 @@ struct SubviewExtents {
   }
 };
 
+template <unsigned TrivialScalarSize>
+struct Padding {
+  // If memory alignment is a multiple of the trivial scalar size then attempt
+  // to align.
+  KOKKOS_INLINE_FUNCTION
```
[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/impl/Kokkos_ViewMapping.hpp:728]: Why?

File: core/src/Kokkos_View.hpp [hunk 28]
```diff
@@ -123,6 +123,32 @@ void runtime_check_rank(const size_t rank, const size_t dyn_rank,
...
+  KOKKOS_INLINE_FUNCTION
+  is_strided_from_padding(const bool&){};
+  KOKKOS_INLINE_FUNCTION
+  bool value() const { return false; }
```
[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/Kokkos_View.hpp:138]: Why?

**Review by masterleinad — COMMENTED** [review]
This looks OK to me but I discussed with @dalg24 that the unit tests should look closer to the ones in https://github.com/kokkos/kokkos/pull/5190/files in the end.

**Comment by janciesko** [comment]
I like the changes in the other PR, I'll compact the tests.

**Comment by masterleinad** [comment]
@janciesko I took the freedom to simplify the containers test some (with the same functionality and adding a test for rank-8 views).

**Comment by janciesko** [comment]
Very good, thanks.

**Comment by PhilMiller** [comment]
Per Daniel, further review and integration is waiting on performance regression testing.

**Comment by ajpowelsnl** [comment]
@janciesko - this is on the list of issues to discuss on Wed.  Are you still good with that?

**Comment by ajpowelsnl** [comment]
- **TODO**
- Integrate MDSpan, and this issue goes away
- MDSpan ETA:  TBD;
- Problem is the layout
- **Decision**
- Fix properly with MDSpan in future
- Don't be sad, @janciesko

**Comment by masterleinad** [comment]
We should just not forget to add tests when this is fixed with `mdspan`.
